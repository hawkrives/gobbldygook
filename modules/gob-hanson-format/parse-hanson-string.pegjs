{
  var globalLastDept

  function storeDept(dept) {
    globalLastDept = dept
  }

  function fetchDept(dept) {
    return globalLastDept
  }

  var flatten = require('lodash/flatten')
  var assign = require('lodash/assign')
}


Result
  = Or


Expression 'expression'
  = _ expr:(
      Not
    / Parenthetical
    / Course
    / Where
    / Occurrence
    / Of
    / Modifier
    / Reference
  ) _
  { return expr }


// Primitives

OptionalS
  = 's'?


Where
  = count:Counter _ distinct:IsDistinct _ 'course' OptionalS _ 'where' _ where:Qualifier
  { return {
      type: 'where',
      count: count,
      qualification: where,
      distinct: distinct,
  } }


Filter
  = 'only' _ distinct:IsDistinct _ 'courses' _ filter:(
      'where' _ where:Qualifier { return {where: where, type: 'FilterWhere'} }
    / 'from' _ ofList:OfCourseList { return {of: ofList, type: 'FilterOf'} }
  )
  { return {type: 'filter', ...filter, distinct: distinct} }


Occurrence
  = count:Counter _ 'occurrence' OptionalS _ 'of' _ course:Course
    { return {
        type: 'occurrence',
        count: count,
        course: course,
    } }


// Primitive Components

Qualifier
  = '{' _ q:OrQualification _ '}'
    { return q }


OrQualification 'qualification-or'
  = lhs:AndQualification _ '|' _ rhs:OrQualification
    { return {
      type: 'BooleanOr',
      values: [lhs, ...(rhs.values || [rhs])],
    } }
  / AndQualification


AndQualification 'qualification-and'
  = lhs:ParentheticalQualification _ '&' _ rhs:AndQualification
    { return {
      type: 'BooleanAnd',
      values: [lhs, ...(rhs.values || [rhs])],
    } }
  / ParentheticalQualification


ParentheticalQualification
  = OpenParen _ q:OrQualification _ CloseParen { return q }
  / Qualification


Qualification
  = key:QualificationField _
    op:Operator _
    value:(
        f:Function _ 'from' _ 'courses' _ 'where' _ q:Qualifier  { return {...f, qualification: q} }
      / word:QualificationValue
      / list:ParentheticalQualificationValue
    )
    { return {
        type: 'Qualification',
        key: key,
        operator: op,
        value: value,
    } }


ParentheticalQualificationValue
  = OpenParen _ value:OrQualificationValue _ CloseParen { return value }


OrQualificationValue
  = lhs:AndQualificationValue _ '|' _ rhs:OrQualificationValue
    { return {
      type: 'BooleanOr',
      values: [lhs, ...(rhs.values || [rhs])],
    } }
  / AndQualificationValue


AndQualificationValue
  = lhs:QualificationValue _ '&' _ rhs:AndQualificationValue
    { return {
      type: 'BooleanAnd',
      values: [lhs, ...(rhs.values || [rhs])],
    } }
  / QualificationValue


QualificationValue
  = num:Integer         { return {type: 'Number', value: num }}
  / word:[a-z0-9_\-]i+  { return {type: 'String', value: word.join('')} }


QualificationField
  = 'gereqs' / 'year' / 'department' / 'level'

Function
  = name:('max' / 'min') _ OpenParen _ prop:QualificationField _ CloseParen
    { return {
      name: name,
      prop: prop,
      type: 'function',
    } }


Operator
  = ('<=')       { return 'Lte' }
  / ('<')        { return 'Lt'  }
  / ('==' / '=') { return 'Eq'  }
  / ('>=')       { return 'Gte' }
  / ('>')        { return 'Gt'  }
  / ('!=')       { return 'Ne'  }


_ 'whitespace'
  = [ \n\t\r]*


Counter
  = count:EnglishInteger             { return { operator: 'Gte', num: count } }
  / 'at most' _ count:EnglishInteger { return { operator: 'Lte', num: count } }
  / 'exactly' _ count:EnglishInteger { return { operator: 'Eq',  num: count } }


EnglishInteger
  = num:(
        '0' / 'zero'
      / '1.5' / 'one-point-five'
      / '1' / 'one'
      / '2' / 'two'
      / '3' / 'three'
      / '4' / 'four'
      / '5' / 'five'
      / '6' / 'six'
      / '7' / 'seven'
      / '8' / 'eight'
      / '9' / 'nine'
      / '10' / 'ten'
    )
    {
           if (num === 'zero')  { return 0 }
      else if (num === 'one')   { return 1 }
      else if (num === 'one-point-five') { return 1.5 }
      else if (num === 'two')   { return 2 }
      else if (num === 'three') { return 3 }
      else if (num === 'four')  { return 4 }
      else if (num === 'five')  { return 5 }
      else if (num === 'six')   { return 6 }
      else if (num === 'seven') { return 7 }
      else if (num === 'eight') { return 8 }
      else if (num === 'nine')  { return 9 }
      else if (num === 'ten')   { return 10 }
      else
        throw new Error(`invalid number "${num}"`)
    }


// Expressions

Not
  = '!' _ value:Expression
    { return {
      type: 'BooleanNot',
      not: value
    } }


Parenthetical
  = OpenParen _ value:Result _ CloseParen
    { return value }

Or
  = lhs:And _ '|' _ rhs:Or
    { return {
      type: 'BooleanOr',
      values: [lhs, ...(rhs.values || [rhs])],
    } }
  / And


And
  = lhs:Expression _ '&' _ rhs:And
    { return {
      type: 'BooleanAnd',
      values: [lhs, ...(rhs.values || [rhs])],
    } }
  / Expression


OfCourseList
  = OpenParen _ ofItems:(
      val:Course
      rest:( _ ',' _ second:Course { return second } )*
      { return [val].concat(rest) }
    )+ _ ','? _ CloseParen
  { return flatten(ofItems) }


OfList
  = OpenParen _ ofItems:(
      val:Result
      rest:( _ ',' _ second:Result { return second } )*
      { return [val].concat(rest) }
    )+ _ ','? _ CloseParen
  { return flatten(ofItems) }


Of
  = count:(
        Counter
      / 'all'  { return { operator: 'Eq', was: 'All' } }
      / 'any'  { return { operator: 'Gte', num: 1, was: 'Any' } }
      / 'none' { return { operator: 'Eq', num: 0, was: 'None' } }
    )
    _ 'of' _ ofList:OfList
    {
      if (count.was === 'all') {
        count.num = ofList.length
      }

      if (ofList.length < count.num) {
        throw new Error(`you requested ${count.num} items, but only gave ${ofList.length} options (${JSON.stringify(ofList)}).`)
      }

      return {
        type: 'Of',
        count: count,
        of: ofList,
      }
    }


ChildList  // select a few requirements to apply the modifier to.
  = OpenParen _ reqs:(
    val:Reference
    rest:( _ ',' _ second:Reference { return second } )*
    { return [val, ...rest] }
  )+ _ ','? _ CloseParen
  { return flatten(reqs) }


Modifier
  = count:Counter _
    what:(
        'course'
      / 'credit'
      / 'department'
      / 'term'
    ) OptionalS _ besides:Besides? _ 'from' _
    from:(
        'children' _ 'where' _ where:Qualifier { return { from: 'ChildrenWhere', qualification: where, children: '$all' } }
      / 'children'                             { return { from: 'Children', children: '$all' }}
      / 'filter' _ 'where' _ where:Qualifier   { return { from: 'FilterWhere', qualification: where }}
      / 'filter'                               { return { from: 'Filter' }}
      / 'courses' _ 'where' _ where:Qualifier  { return { from: 'Where', qualification: where } }
      / c:ChildList _ 'where' _ w:Qualifier    { return { from: 'ChildrenWhere', qualification: w, children: c } }
      / children:ChildList                     { return { from: 'Children', children: children} }  // an alternative to "from [all] children"
      / child:Reference                        { return { from: 'Children', children: [child]} }
    )
    {
      if (from.from === 'where' && what === 'department') {
        throw new Error('cannot use a modifier with "departments" or "department"')
      }
      if (from.from === 'where' && what === 'term') {
        throw new Error('cannot use a modifier with "terms" or "term"')
      }
      if (from.from === 'children-where' && what !== 'course') {
        throw new Error('must use "courses from" with "children where"')
      }
      if (count.operator !== 'Gte' && what !== 'course') {
        throw new Error('can only use at-least style counters with non-course requests')
      }

      what = what[0].toUpperCase() + what.slice(1)

      let result = {
        ...from,
        type: 'Modifier',
        count: count,
        what: what,
      }
      if (besides) {
        result.besides = besides
      }
      return result
    }


Besides
  = 'besides' _ course:Course { return course }


RequirementTitle
  = title:(
      initial:[A-Z0-9]
      rest:[A-Za-z0-9_\- /'.]*
      { return initial + rest.join('') }
    )
    { return title.trim() }


Reference 'requirement reference'
  = title:(
      a:RequirementTitle
      b:(_ OpenParen t:RequirementTitle CloseParen { return ` (${t})` })?
      { return `${a}${b || ''}` }
    )
    {
      if (options.abbreviations && title in options.abbreviations) {
        title = options.abbreviations[title]
      }
      else if (options.titles && title in options.titles) {
        title = options.titles[title]
      }
      return {
        type: 'Reference',
        requirement: title,
      }
    }


// Course

Course
  = dept:CourseDepartment? _
    num:CourseNumber
    details:(
      '.' year:CourseYear sub:(
        '.' semester:CourseSemester { return {semester} }
      )? { return assign({}, sub, {year}) }
    )?
  {
    return {
      type: 'Course',
      ...details,
      ...(dept || fetchDept()),
      ...num,
    }
  }


CourseDepartment
  = dept:[A-Z/]+
    {
      let department = {department: dept.join('').split('/')}
      storeDept(department)
      return department
    }


CourseNumber 'course number'
  = nums:(Digit Digit Digit)
    international:'I'? lab:'L'?
    {
      let result = {}

      let number = parseInt(nums.join(''))

      if (international) {
        result.international = true
      }
      if (lab) {
        result.type = 'Lab'
      }

      return assign({}, result, {number})
    }


CourseYear
  = nums:(Digit Digit Digit Digit)
    { return parseInt(nums.join('')) }
  / Asterisk
  / Else { throw new SyntaxError('A course year must be either a four-digit year [e.g. 1994] or an asterisk [*].')}


CourseSemester
  = num:[1-5] { return parseInt(num) }
  / Asterisk
  / Else { throw new SyntaxError('A course semester must be either a number between 1 [Fall] and 5 [Summer Session 2], or an asterisk [*].')}


// Primitives

Else
  = .+


UppercaseLetter
  = char:[A-Z]
    { return char }


Asterisk
  = '*'


Word
  = chars:[a-z]i+
    { return chars.join('') }


Integer
  = digits:Digit+
    { return parseInt(digits.join('')) }


Digit
  = num:[0-9]
    { return parseInt(num) }


OpenParen
  = '('


CloseParen
  = ')'


IsDistinct
  = d:'distinct'?
  { return Boolean(d) }
