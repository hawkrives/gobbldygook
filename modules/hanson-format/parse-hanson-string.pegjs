{
  let globalLastDept

  function storeDept(dept) {
    globalLastDept = dept
  }

  function fetchDept(dept) {
    return globalLastDept
  }

  const flatten = require('lodash/flatten')
  const assign = require('lodash/assign')
  let normalizeDepartment
  try {
    normalizeDepartment = require('./convert-department').normalizeDepartment
  } catch (e) {
    normalizeDepartment = x => x
  }
}


Result
  = Or


Expression 'expression'
  = _ expr:(
      Not
    / Parenthetical
    / Course
    / Where
    / Occurrence
    / Of
    / Modifier
    / Reference
  ) _
  { return expr }


// Primitives

OptionalS
  = 's'?


Where
  = count:Counter _ distinct:IsDistinct _ 'course' OptionalS _ 'where' _ where:Qualifier
  { return {
      $type: 'where',
      $count: count,
      $where: where,
      $distinct: distinct,
  } }


Filter
  = 'only' _ distinct:IsDistinct _ 'courses' _ filter:(
      'where' _ where:Qualifier { return {$where: where, $filterType: 'where'} }
    / 'from' _ ofList:OfList { return {$of: ofList, $filterType: 'of'} }
  )
  { return assign({}, filter, {$distinct: distinct, $type: 'filter'}) }


Occurrence
  = count:Counter _ 'occurrence' OptionalS _ 'of' _ course:Course
    { return {
        $type: 'occurrence',
        $count: count,
        $course: course.$course,
    } }


// Primitive Components

Qualifier
  = '{' _ q:OrQualification _ '}'
    { return q }


OrQualification 'qualification-or'
  = lhs:AndQualification _ '|' _ rhs:OrQualification
    { return {
      $type: 'boolean',
      $booleanType: 'or',
      $or: [lhs].concat('$or' in rhs ? rhs.$or : [rhs]),
    } }
  / AndQualification


AndQualification 'qualification-and'
  = lhs:ParentheticalQualification _ '&' _ rhs:AndQualification
    { return {
      $type: 'boolean',
      $booleanType: 'and',
      $and: [lhs].concat('$and' in rhs ? rhs.$and : [rhs]),
    } }
  / ParentheticalQualification


ParentheticalQualification
  = OpenParen _ q:OrQualification _ CloseParen { return q }
  / Qualification


Qualification
  = key:Word _
    op:Operator _
    value:(
        f:Function _ 'from' _ 'courses' _ 'where' _ q:Qualifier  { return assign({}, f, {$where: q}) }
      / word:QualificationValue
      / list:ParentheticalQualificationValue
    )
    { return {
        $type: 'qualification',
        $key: key,
        $operator: op,
        $value: value,
    } }


ParentheticalQualificationValue
  = OpenParen _ value:OrQualificationValue _ CloseParen { return value }


OrQualificationValue
  = lhs:AndQualificationValue _ '|' _ rhs:OrQualificationValue
    { return {
      $type: 'boolean',
      $booleanType: 'or',
      $or: [lhs].concat(rhs.$or ? rhs.$or : [rhs]),
    } }
  / AndQualificationValue


AndQualificationValue
  = lhs:QualificationValue _ '&' _ rhs:AndQualificationValue
    { return {
      $type: 'boolean',
      $booleanType: 'and',
      $and: [lhs].concat(rhs.$and ? rhs.$and : [rhs]),
    } }
  / QualificationValue


QualificationValue
  = num:Integer         { return num }
  / word:[a-z0-9_\-]i+  { return word.join('') }


Function
  = name:Word _ OpenParen _ prop:Word _ CloseParen
    { return {
      $name: name,
      $prop: prop,
      $type: 'function',
    } }


Operator
  = ('<=')       { return '$lte' }
  / ('<')        { return '$lt'  }
  / ('==' / '=') { return '$eq'  }
  / ('>=')       { return '$gte' }
  / ('>')        { return '$gt'  }
  / ('!=')       { return '$ne'  }


_ 'whitespace'
  = [ \n\t\r]*


Counter
  = count:EnglishInteger             { return { $operator: '$gte', $num: count } }
  / 'at most' _ count:EnglishInteger { return { $operator: '$lte', $num: count } }
  / 'exactly' _ count:EnglishInteger { return { $operator: '$eq',  $num: count } }


EnglishInteger
  = num:(
        'zero'
      / 'one-point-five'
      / 'one'
      / 'two'
      / 'three'
      / 'four'
      / 'five'
      / 'six'
      / 'seven'
      / 'eight'
      / 'nine'
      / 'ten'
    )
    {
           if (num === 'zero')  { return 0 }
      else if (num === 'one')   { return 1 }
      else if (num === 'one-point-five') { return 1.5 }
      else if (num === 'two')   { return 2 }
      else if (num === 'three') { return 3 }
      else if (num === 'four')  { return 4 }
      else if (num === 'five')  { return 5 }
      else if (num === 'six')   { return 6 }
      else if (num === 'seven') { return 7 }
      else if (num === 'eight') { return 8 }
      else if (num === 'nine')  { return 9 }
      else if (num === 'ten')   { return 10 }
      else
        throw new Error(`invalid number "${num}"`)
    }


// Expressions

Not
  = '!' _ value:Expression
    { return {
      $type: 'boolean',
      $not: value
    } }


Parenthetical
  = OpenParen _ value:Result _ CloseParen
    { return value }

Or
  = lhs:And _ '|' _ rhs:Or
    { return {
      $type: 'boolean',
      $booleanType: 'or',
      $or: [lhs].concat('$or' in rhs ? rhs.$or : [rhs]),
    } }
  / And


And
  = lhs:Expression _ '&' _ rhs:And
    { return {
      $type: 'boolean',
      $booleanType: 'and',
      $and: [lhs].concat('$and' in rhs ? rhs.$and : [rhs]),
    } }
  / Expression


OfList
  = OpenParen _ ofItems:(
      val:Result
      rest:( _ ',' _ second:Result { return second } )*
      { return [val].concat(rest) }
    )+ _ ','? _ CloseParen
  { return flatten(ofItems) }


Of
  = count:(
        Counter
      / 'all'  { return { $operator: '$eq', $was: 'all' } }
      / 'any'  { return { $operator: '$gte', $num: 1, $was: 'any' } }
      / 'none' { return { $operator: '$eq', $num: 0, $was: 'none' } }
    )
    _ 'of' _ ofList:OfList
    {
      if (count.$was === 'all') {
        count.$num = ofList.length
      }

      if (ofList.length < count.$num) {
        throw new Error(`you requested ${count.$num} items, but only gave ${ofList.length} options (${JSON.stringify(ofList)}).`)
      }

      return {
        $type: 'of',
        $count: count,
        $of: ofList,
      }
    }


ChildList  // select a few requirements to apply the modifier to.
  = OpenParen _ reqs:(
    val:Reference
    rest:( _ ',' _ second:Reference { return second } )*
    { return [val].concat(rest) }
  )+ _ ','? _ CloseParen
  { return flatten(reqs) }


Modifier
  = count:Counter _
    what:(
        'course'
      / 'credit'
      / 'department'
    ) OptionalS _ besides:Besides? _ 'from' _
    from:(
        'children' _ 'where' _ where:Qualifier { return { $from: 'children-where', $where: where, $children: '$all' } }
      / 'children'                             { return { $from: 'children', $children: '$all' }}
      / 'filter' _ 'where' _ where:Qualifier   { return { $from: 'filter-where', $where: where }}
      / 'filter'                               { return { $from: 'filter' }}
      / 'courses' _ 'where' _ where:Qualifier  { return { $from: 'where', $where: where } }
      / c:ChildList _ 'where' _ w:Qualifier    { return { $from: 'children-where', $where: w, $children: c } }
      / children:ChildList                     { return { $from: 'children', $children: children} }  // an alternative to "from [all] children"
      / child:Reference                        { return { $from: 'children', $children: [child]} }
    )
    {
      if (from.$from === 'where' && what === 'department') {
        throw new Error('cannot use a modifier with "departments" or "department"')
      }
      if (from.$from === 'children-where' && what !== 'course') {
        throw new Error('must use "courses from" with "children where"')
      }
      if (count.$operator !== '$gte' && what !== 'course') {
        throw new Error('can only use at-least style counters with non-course requests')
      }
      let result = assign({}, from, {
        $type: 'modifier',
        $count: count,
        $what: what,
      })
      if (besides) {
        result.$besides = besides
      }
      return result
    }


Besides
  = 'besides' _ course:Course { return course }


RequirementTitle
  = title:(
      initial:[A-Z0-9]
      rest:[A-Za-z0-9_\- /'.]*
      { return initial + rest.join('') }
    )
    { return title.trim() }


Reference 'requirement reference'
  = title:(
      a:RequirementTitle
      b:(_ OpenParen t:RequirementTitle CloseParen { return ` (${t})` })?
      { return `${a}${b || ''}` }
    )
    {
      if (options.abbreviations && title in options.abbreviations) {
        title = options.abbreviations[title]
      }
      else if (options.titles && title in options.titles) {
        title = options.titles[title]
      }
      return {
        $type: 'reference',
        $requirement: title,
      }
    }


// Course

Course
  = dept:CourseDepartment? _
    num:CourseNumber
    details:(
      '.' section:CourseSection sub:(
        '.' year:CourseYear sub:(
          '.' semester:CourseSemester { return {semester} }
        )? { return assign({}, sub, {year}) }
      )? { return assign({}, sub, {section}) }
    )?
  {
    return {
      $type: 'course',
      $course: assign({},
        details,
        (dept || fetchDept()),
        num
      ),
    }
  }


CourseDepartment
  = dept1:(c1:UppercaseLetter c2:UppercaseLetter { return c1 + c2 })
    part2:(
      '/' l1:UppercaseLetter l2:UppercaseLetter
          { return {dept: l1 + l2, type: 'separate'} }
      / chars:UppercaseLetter*
          { return {dept: chars.join(''), type: 'joined'} }
    )
    {
      const {type, dept: dept2} = part2
      let department
      if (type === 'joined') {
        department = {department: [dept1 + dept2]}
      }
      else if (type === 'separate') {
        department = {department: [
          normalizeDepartment(dept1),
          normalizeDepartment(dept2),
        ]}
      }
      storeDept(department)
      return department
    }


CourseNumber 'course number'
  = nums:(Digit Digit Digit)
    international:'I'? lab:'L'?
    {
      let result = {}

      let number = parseInt(nums.join(''))

      if (international) {
        result.international = true
      }
      if (lab) {
        result.type = 'Lab'
      }

      return assign({}, result, {number})
    }


CourseSection
  = UppercaseLetter
  / Asterisk
  / Else { throw new SyntaxError('A course section must be either an uppercase letter [A-Z] or an asterisk [*].')}


CourseYear
  = nums:(Digit Digit Digit Digit)
    { return parseInt(nums.join('')) }
  / Asterisk
  / Else { throw new SyntaxError('A course year must be either a four-digit year [e.g. 1994] or an asterisk [*].')}


CourseSemester
  = num:[1-5] { return parseInt(num) }
  / Asterisk
  / Else { throw new SyntaxError('A course semester must be either a number between 1 [Fall] and 5 [Summer Session 2], or an asterisk [*].')}


// Primitives

Else
  = .+


UppercaseLetter
  = char:[A-Z]
    { return char }


Asterisk
  = '*'


Word
  = chars:[a-z]i+
    { return chars.join('') }


Integer
  = digits:Digit+
    { return parseInt(digits.join('')) }


Digit
  = num:[0-9]
    { return parseInt(num) }


OpenParen
  = '('


CloseParen
  = ')'


IsDistinct
  = d:'distinct'?
  { return Boolean(d) }
