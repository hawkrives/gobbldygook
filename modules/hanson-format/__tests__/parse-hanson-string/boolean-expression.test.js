import { customParser } from './parse-hanson-string.support'
const parse = customParser({ allowedStartRules: ['Or'] })

describe('BooleanExpression', () => {
    it('parses courses separated by | as being or-d', () => {
        expect(parse('CSCI 121 | CSCI 125')).toMatchSnapshot()
    })

    it('parses courses separated by & as being and-d', () => {
        expect(parse('CSCI 121 & CSCI 125')).toMatchSnapshot()
    })

    it('parses courses with no departments after an prior department', () => {
        expect(parse('CSCI 121 | 125')).toMatchSnapshot()
    })

    it('changes departments when given a new one', () => {
        expect(parse('CSCI 121 | PSCI 125')).toMatchSnapshot()
    })

    it('allows several &-d courses in a row', () => {
        expect(parse('CSCI 121 & 125 & 126 & 123')).toMatchSnapshot()
    })

    it('allows several |-d courses in a row', () => {
        expect(parse('CSCI 121 | 125 | 126 | 123')).toMatchSnapshot()
    })

    it('keeps duplicates in a list of courses', () => {
        expect(parse('CSCI 121 | 121 | 125')).toMatchSnapshot()
    })

    it('allows a & b | c – boolean logic for courses', () => {
        expect(parse('CSCI 121 & 122 | 123')).toMatchSnapshot()
    })

    it('allows a | b & c – boolean logic for courses', () => {
        expect(parse('CSCI 121 | 122 & 123')).toMatchSnapshot()
    })

    it('supports parentheses to control order-of-operations - (a | b) & c', () => {
        expect(parse('(CSCI 121 | 122) & 123')).toMatchSnapshot()
    })
})
