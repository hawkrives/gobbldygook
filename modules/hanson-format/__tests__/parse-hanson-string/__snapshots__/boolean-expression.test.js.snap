// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BooleanExpression allows a & b | c – boolean logic for courses 1`] = `
Object {
  "$booleanType": "or",
  "$or": Array [
    Object {
      "$and": Array [
        Object {
          "$course": Object {
            "department": Array [
              "CSCI",
            ],
            "number": 121,
          },
          "$type": "course",
        },
        Object {
          "$course": Object {
            "department": Array [
              "CSCI",
            ],
            "number": 122,
          },
          "$type": "course",
        },
      ],
      "$booleanType": "and",
      "$type": "boolean",
    },
    Object {
      "$course": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 123,
      },
      "$type": "course",
    },
  ],
  "$type": "boolean",
}
`;

exports[`BooleanExpression allows a | b & c – boolean logic for courses 1`] = `
Object {
  "$booleanType": "or",
  "$or": Array [
    Object {
      "$course": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 121,
      },
      "$type": "course",
    },
    Object {
      "$and": Array [
        Object {
          "$course": Object {
            "department": Array [
              "CSCI",
            ],
            "number": 122,
          },
          "$type": "course",
        },
        Object {
          "$course": Object {
            "department": Array [
              "CSCI",
            ],
            "number": 123,
          },
          "$type": "course",
        },
      ],
      "$booleanType": "and",
      "$type": "boolean",
    },
  ],
  "$type": "boolean",
}
`;

exports[`BooleanExpression allows several &-d courses in a row 1`] = `
Object {
  "$and": Array [
    Object {
      "$course": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 121,
      },
      "$type": "course",
    },
    Object {
      "$course": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 125,
      },
      "$type": "course",
    },
    Object {
      "$course": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 126,
      },
      "$type": "course",
    },
    Object {
      "$course": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 123,
      },
      "$type": "course",
    },
  ],
  "$booleanType": "and",
  "$type": "boolean",
}
`;

exports[`BooleanExpression allows several |-d courses in a row 1`] = `
Object {
  "$booleanType": "or",
  "$or": Array [
    Object {
      "$course": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 121,
      },
      "$type": "course",
    },
    Object {
      "$course": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 125,
      },
      "$type": "course",
    },
    Object {
      "$course": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 126,
      },
      "$type": "course",
    },
    Object {
      "$course": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 123,
      },
      "$type": "course",
    },
  ],
  "$type": "boolean",
}
`;

exports[`BooleanExpression changes departments when given a new one 1`] = `
Object {
  "$booleanType": "or",
  "$or": Array [
    Object {
      "$course": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 121,
      },
      "$type": "course",
    },
    Object {
      "$course": Object {
        "department": Array [
          "PSCI",
        ],
        "number": 125,
      },
      "$type": "course",
    },
  ],
  "$type": "boolean",
}
`;

exports[`BooleanExpression keeps duplicates in a list of courses 1`] = `
Object {
  "$booleanType": "or",
  "$or": Array [
    Object {
      "$course": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 121,
      },
      "$type": "course",
    },
    Object {
      "$course": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 121,
      },
      "$type": "course",
    },
    Object {
      "$course": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 125,
      },
      "$type": "course",
    },
  ],
  "$type": "boolean",
}
`;

exports[`BooleanExpression parses courses separated by & as being and-d 1`] = `
Object {
  "$and": Array [
    Object {
      "$course": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 121,
      },
      "$type": "course",
    },
    Object {
      "$course": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 125,
      },
      "$type": "course",
    },
  ],
  "$booleanType": "and",
  "$type": "boolean",
}
`;

exports[`BooleanExpression parses courses separated by | as being or-d 1`] = `
Object {
  "$booleanType": "or",
  "$or": Array [
    Object {
      "$course": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 121,
      },
      "$type": "course",
    },
    Object {
      "$course": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 125,
      },
      "$type": "course",
    },
  ],
  "$type": "boolean",
}
`;

exports[`BooleanExpression parses courses with no departments after an prior department 1`] = `
Object {
  "$booleanType": "or",
  "$or": Array [
    Object {
      "$course": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 121,
      },
      "$type": "course",
    },
    Object {
      "$course": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 125,
      },
      "$type": "course",
    },
  ],
  "$type": "boolean",
}
`;

exports[`BooleanExpression supports parentheses to control order-of-operations - (a | b) & c 1`] = `
Object {
  "$and": Array [
    Object {
      "$booleanType": "or",
      "$or": Array [
        Object {
          "$course": Object {
            "department": Array [
              "CSCI",
            ],
            "number": 121,
          },
          "$type": "course",
        },
        Object {
          "$course": Object {
            "department": Array [
              "CSCI",
            ],
            "number": 122,
          },
          "$type": "course",
        },
      ],
      "$type": "boolean",
    },
    Object {
      "$course": Object {
        "department": Array [
          "CSCI",
        ],
        "number": 123,
      },
      "$type": "course",
    },
  ],
  "$booleanType": "and",
  "$type": "boolean",
}
`;
